doc/compiler/osr/OSR.md

```diff
@@ -0,0 +1,276 @@
+                  \          |
+                    --------------------------
+                    |                        |
+                    --------------------------
+                             |
+                    --------------------------    ---------------------
+                    |     exit               |    | catch 1           |
+                    --------------------------    ---------------------
+                                                          |
+                                                  ---------------------
+                                                  |prepareForOSRHelper|___>  exit
+                                                  |   iconst -1       |
+                                                  ---------------------
+
+
+```
+
+
+## Supporting Classes
+
+**OSRGuardAnalysis**  : detects OSR guards that are no longer required
+
+**OSRGuardInsertion** :  replaces HCR guards with OSR guards based on the fear analysis described above.
+
+**OSRGuardRemoval** : removes OSR guards that became redundant due to optimizations after OSRGuardInsertion. Uses OSRGuardAnalysis
+
+
+## nextGenHCR Control Options
+
+As this feature relies on HCR and OSR, both will be enabled when in nextGenHCR.
+To prevent these from being enabled and to avoid any additional modifications to the trees, use -Xjit:disableNextGenHCR.
+In the event that disableOSR or the environment variable TR_DisableHCR is set, nextGenHCR will disable itself.
+It should be noted that if the VM requests HCR when nextGenHCR is disabled, the traditional implementation will be used.
+Setting the environment variable TR_DisableHCRGuards will result in nextGenHCR working as described above,
+but it will not insert any OSR guards as there will be no HCR guards to replace.
+This is a debug mode to evaluate OSR overhead and it will not correctly implement redefinitions.
+There is a similar TR_DisableOSRGuards variable, which will prevent nextGenHCR from removing HCR guards and inserting OSR guards,
+but will still include the other OSR and HCR overheads.
+
+Both HCR implementations will be disabled under FSD, as HCR implements a subset of its functionality.
+Moreover, nextGenHCR requires VM support with a feature described by the flag OSRSafePoint.
+If this is not enabled, nextGenHCR will disable itself.
+Finally, there are a number of situations where we will fallback on the traditional HCR implementation to reduce compile time overhead.
+These include cold compiles, profiling compiles and DLT.
+
+
+# Static Final Field Folding
+
+This optimization uses the same infrastructure as nextGenHCR except that the fear points are the points just before the initial value of a static final field being used.
+
+
+# OSR On Guard Failure
+
+When OSR on guard failure is enabled (with -Xjit:enableOSROnGuardFailure) when a guard fails, instead of taking the cold path we trigger an OSR. This is implemented as a voluntary pre-execution OSR.
```

Comment:
Maybe instead of "When OSR on guard failure is enabled" this should say "If OSR on guard failure is enabled. . . . "